pixl_dcmd/src/pixl_dcmd/main.py

```diff
@@ -155,292 +131,130 @@ def remove_overlays(dataset: Dataset) -> Dataset:
-
-                dataset[grp, el].value = hashed_value
-
-                message = f"Changing: {name} (0x{grp:04x},0x{el:04x})"
-                logger.debug(f"\t{message}")
-            else:
-                message = f"Missing: {name} (0x{grp:04x},0x{el:04x})\
-                 - Operation ({op})"
-                logger.warning(f"\t{message}")
 
-    return dataset
+def _secure_hash(dataset: Dataset, tag: tuple, mrn: str, accession_number: str) -> None:
+    """
+    Use the hasher API to consistently but securely hash ids later used for linking.
+    """
+    grp = tag[0]
+    el = tag[1]
+
+    if tag in dataset:
+        message = f"Securely hashing: (0x{grp:04x},0x{el:04x})"
+        logger.debug(f"\t{message}")
+        if grp == 0x0010 and el == 0x0020:  # Patient ID
+            pat_value = mrn + accession_number
 
+            hashed_value = _hash_values(pat_value)
+            # Query PIXL database
+            existing_image = query_db(mrn, accession_number)
+            # Insert the hashed_value into the PIXL database
+            add_hashed_identifier_and_save(existing_image, hashed_value)
+        elif dataset[grp, el].VR == "SH":
+            pat_value = str(dataset[grp, el].value)
+            hashed_value = _hash_values(pat_value, 16)
+            # Do we want to store other hashed values in the db?
 
-def hash_endpoint_path_for_tag(group: bytes, element: bytes) -> str:
-    """Call a hasher endpoint depending on the dicom tag group and emement"""
-    if group == 0x0010 and element == 0x0020:  # Patient ID
-        return "/hash-mrn"
-    if group == 0x0008 and element == 0x0050:  # Accession Number
-        return "/hash-accession-number"
+        dataset[grp, el].value = hashed_value
 
-    return "/hash"
+    else:
+        message = f"Missing linking variable (0x{grp:04x},0x{el:04x})"
+        logger.warning(f"\t{message}")
 
 
-def _hash_values(grp: bytes, el: bytes, pat_value: str, hasher_host_url: str) -> str:
-    ep_path = hash_endpoint_path_for_tag(group=grp, element=el)
-    payload = ep_path + "?message=" + pat_value
-    request_url = hasher_host_url + payload
-    response = requests.get(request_url)
-    # All three hashing endpoints return application/text so should be fine to
-    # use response.text here
+def _hash_values(pat_value: str, hash_len: int = 0) -> str:
+    """
+    Utility function for hashing values using the hasher API.
+    """
+    HASHER_API_AZ_NAME = config("HASHER_API_AZ_NAME")
+    HASHER_API_PORT = config("HASHER_API_PORT")
+    hasher_req_url = (
+        f"http://{HASHER_API_AZ_NAME}:{HASHER_API_PORT}/hash?message={pat_value}"
+    )
+    if hash_len:
+        hasher_req_url += f"&length={hash_len}"
+    response = requests.get(hasher_req_url)
     logger.debug("RESPONSE = %s}" % response.text)
     return response.text
+
+
+def enforce_whitelist(dataset: Dataset, tags: dict[tuple, str]) -> Dataset:
+    """Delete any tags not in the tagging scheme."""
+    # For every element:
```

Comment:
so... I added this test and it fails, good catch (see latest commit in this pr). 

This wasn't part of our original solution, nor the considerations when picking the library, so I guess that's not surprising?

Nevertheless I tried to make it work in several ways (with recursion, all_elements iteration, etc), but I was not successful, see latest commit on [this branch](https://github.com/UCLH-Foundry/PIXL/tree/peshence/291-sq-mess) - it doesn't reflect everything I tried, just where i gave up. 

It might still be possible but it's getting to the point where we're not really using the library, tbh.

Reply:
Yeah before when we were dropping all private tags this was done for us by pydicom, so it slipped off the radar. I wonder if its worth writing a custom function using `walk` that just removes all sequences?

Reply:
Ah, yeah, doesn't look like that the Kitware dicom-anonymiser directly handles sequences, and it wasn't a feature I was specifically looking out for in my options analysis. Each sequence (VR=SQ) is a DICOM dataset in its own right, so you could use pydicom to select it, and then use the same anonymisation approach for it, and then place it back into the main dataset?

(Looking at some of the other options I looked at, [the Flywheel Migration Toolkit allows direct addressing of sequence elements](https://flywheel-io.gitlab.io/public/migration-toolkit/pages/file_profiles.html#fields) and [deid flattens sequences out](https://pydicom.github.io/deid/examples/func-sequence-replace/))